### PR TITLE
Remove boilerplate in favour of Attributes

### LINKED ISSUE #35: Refactor Checks to reduce boilerplate
#### Description:

- Use attributes to filter checks based on file path pattern (rather than by namespace)
- Make each check method only perform one discreet compliance check and only return a span or null
- Use an attribute to provide the error message (possibly tokenised to allow for dynamic content)

### PR DESCRIPTION
**Pull request by awakeman**
### Issues addressed by this PR

 Closes #35 

 
For starters Checks methods have been modified to simply return a span or null for fail and pass respectively. Attributes have been added to supply an error message (currently that means messages are not dynamic) and severity for if it fails. Attributes have also been added to select the tests to run for each part of the syntax tree. Primarily based on path pattern matching but also for filtering out private methods (for example).

This means that the applicability of a check is no longer defined by the namespace it is in but its place in the folder hierarchy, meaning a namespace check can fail without hiding method/class checks for that type of method.

Probably worth wrapping some of the path stuff into attributes that encapsulate the matches for various defined categories (such as Engine Create methods or BHoM classes) do reduce duplication of patterns.

 ### Test files

 ### Changelog

 ### Additional comments

### CODE REVIEW

File: Test_Engine/Query/Checks/InputAttributeHasParameter.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        [Message("Input attribute does not match any parameter")]
+        [ErrorLevel(ErrorLevel.Warning)]
+        [Path(@"([a-zA-Z0-9]+)_(Engine|Adapter)\\.*\.cs$")]
+        public static Span InputAttributeHasParameter(AttributeSyntax node)
+        {
+            if (node.Name.ToString() != "Input") return null;
+
+            var method = node.Parent.Parent as BaseMethodDeclarationSyntax;
+            if (method != null && method.IsPublic() && (method.IsEngineMethod() || method.IsAdapterConstructor()))
+            {
```
Comment (awakeman) [inline Test_Engine/Query/Checks/InputAttributeHasParameter.cs]: @FraserGreenroyd this here is still boilerplate, need to think if there is a nice way to handle this kind of stuff without it, I have some ideas.

Comment (FraserGreenroyd) [reply Test_Engine/Query/Checks/InputAttributeHasParameter.cs]: We can continue to refactor during the week, I'm enthused by what you've done here thus far and what we've currently got - going to start using it on some toolkits for testing purposes by the end of the week to see what else we can achieve.

**Comment by FraserGreenroyd** [comment]
Reminder to double check for morning: do convert methods run through the `MethodNameMustMatchFileName` check? If so, this should be avoided as the `IsValidConvertMethodName` will check valid Convert method names

**Comment by FraserGreenroyd** [comment]
Ok so checked today instead and a fix is pushed.

**Pull request merged**